### PR TITLE
WebAPI: Echobot with Counter

### PR DESCRIPTION
**Pull request by daveta**
I think we can independently review the code here.  
But this issue was ran into:
https://github.com/Microsoft/botbuilder-dotnet/issues/920

I think we should pull Drew into this one as well.

### CODE REVIEW

File: csharp_webapi/2.EchoBot-With-Counter/EchoBot.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+
+namespace EchoBotWithCounter
+{
+    /// <summary>
+    /// Represents a bot that processes incoming activities.
+    /// For each user interaction, an instance of this class is created and the OnTurnAsync method is called.
+    /// This is a Transient lifetime service.  Transient lifetime services are created
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/EchoBot.cs]: >This is a Transient lifetime service [](start = 8, length = 36)

I suspect the discussions around "Transient lifetime service" are *only* valid for ASP.Net Core 2. Consider replacing with the WebAPI relevant text around lifecycle management.

File: csharp_webapi/2.EchoBot-With-Counter/EchoBot.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        /// <seealso cref="BotStateSet"/>
+        /// <seealso cref="ConversationState"/>
+        /// <seealso cref="IMiddleware"/>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (turnContext.Activity.Type == ActivityTypes.Message)
+            {
+                // Get the conversation state from the turn context.
+                var state = await _accessors.CounterState.GetAsync(turnContext, () => new CounterState());
+
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/EchoBot.cs]: Consider adding a comment around the lambda that is setting the default value. I *suspect* we can remove this altogether and put the logic in the POCO.

File: csharp_webapi/2.EchoBot-With-Counter/EchoBot.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                // Bump the turn count for this conversation.
+                state.TurnCount++;
+
+                // Save the new turn count into the conversation state.
+                await _accessors.ConversationState.SaveChangesAsync(turnContext);
+
+                // Echo back to the user whatever they typed.
+                var responseMessage = $"Turn {state.TurnCount}: You sent '{turnContext.Activity.Text}'\n";
+                await turnContext.SendActivityAsync(responseMessage);
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/EchoBot.cs]: is "\n" needed at the end? I wouldn't think so.

File: csharp_webapi/2.EchoBot-With-Counter/EchoBot.csproj [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+      <DependentUpon>Web.config</DependentUpon>
+    </None>
+    <None Include="Web.Release.config">
+      <DependentUpon>Web.config</DependentUpon>
+    </None>
+  </ItemGroup>
+  <ItemGroup>
+    <Analyzer Include="packages\AsyncUsageAnalyzers.1.0.0-alpha003\analyzers\dotnet\AsyncUsageAnalyzers.dll" />
+    <Analyzer Include="packages\StyleCop.Analyzers.1.1.0-beta008\analyzers\dotnet\cs\StyleCop.Analyzers.CodeFixes.dll" />
+    <Analyzer Include="packages\StyleCop.Analyzers.1.1.0-beta008\analyzers\dotnet\cs\StyleCop.Analyzers.dll" />
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/EchoBot.csproj]: This one isn't needed. I think.

File: csharp_webapi/2.EchoBot-With-Counter/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+﻿This sample demonstrates a simple echo bot using state with ASP.Net Core 2.
+
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/README.md]: Using WebAPI, not Core2

File: csharp_webapi/2.EchoBot-With-Counter/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+- Bring up a terminal, navigate to `BotBuilder-Samples\csharp_webapi\2.EchoBot-With-Counter` folder.
+- Type 'dotnet run'.
+
+## Testing the bot using Bot Framework Emulator
+[Microsoft Bot Framework Emulator](https://github.com/microsoft/botframework-emulator) is a desktop application that allows bot 
+developers to test and debug their bots on localhost or running remotely through a tunnel.
+- Install the Bot Framework emulator from [here](https://aka.ms/botframeworkemulator).
+
+## Connect to bot using Bot Framework Emulator **V4**
+- Launch the Bot Framework Emulator.
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/README.md]: Replace **V4** with just V4

File: csharp_webapi/2.EchoBot-With-Counter/App_Start/BotConfig.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+using Microsoft.Bot.Builder.Integration.AspNet.WebApi;
+using Microsoft.Bot.Configuration;
+using Unity;
+using Unity.Lifetime;
+
+namespace EchoBotWithCounter
+{
+    /// <summary>
+    /// Performs the Bot-specific configuration during Asp.Net start.
+    /// </summary>
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/App_Start/BotConfig.cs]: bot-specific. Or even bot specific.

File: csharp_webapi/2.EchoBot-With-Counter/App_Start/BotServices.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        /// <param name="luisServices">A dictionary of named <see cref="LuisRecognizer"/> instances for usage within the bot.</param>
+        public BotServices(Dictionary<string, LuisRecognizer> luisServices)
+        {
+            LuisServices = luisServices ?? throw new ArgumentNullException(nameof(luisServices));
+        }
+
+        /// <summary>
+        /// Gets the set of LUIS Services used.
+        /// Given there can be multiple <see cref="LuisRecognizer"/> services used in a single bot,
+        /// LuisServices is represented as a dictionary.  This is also modeled in the
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/App_Start/BotServices.cs]: >   [](start = 56, length = 2)

Extra space after "."

File: csharp_webapi/2.EchoBot-With-Counter/App_Start/UnityConfig.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        private static Lazy<IUnityContainer> container =
+          new Lazy<IUnityContainer>(() =>
+          {
+              var container = new UnityContainer();
+              RegisterTypes(container);
+              return container;
+          });
+
+        /// <summary>
+        /// Gets configured Unity Container.
```
Comment (cleemullins) [inline csharp_webapi/2.EchoBot-With-Counter/App_Start/UnityConfig.cs]: >ts co [](start = 14, length = 5)

Gets the configured

File: csharp_webapi/2.EchoBot-With-Counter/EchoBot.csproj (hunk 4 above)
Comment (daveta) [reply csharp_webapi/2.EchoBot-With-Counter/EchoBot.csproj]: @cleemullins : This is still left in for fear of breaking fragile webapi UI.

File: csharp_webapi/2.EchoBot-With-Counter/README.md (hunk 5 above)
Comment (daveta) [reply csharp_webapi/2.EchoBot-With-Counter/README.md]: Nice

File: csharp_webapi/2.EchoBot-With-Counter/README.md (hunk 6 above)
Comment (daveta) [reply csharp_webapi/2.EchoBot-With-Counter/README.md]: Fixed

File: csharp_webapi/2.EchoBot-With-Counter/App_Start/BotConfig.cs (hunk 7 above)
Comment (daveta) [reply csharp_webapi/2.EchoBot-With-Counter/App_Start/BotConfig.cs]: Cool

File: csharp_webapi/2.EchoBot-With-Counter/App_Start/BotServices.cs (hunk 8 above)
Comment (daveta) [reply csharp_webapi/2.EchoBot-With-Counter/App_Start/BotServices.cs]: Nice

File: csharp_webapi/2.EchoBot-With-Counter/App_Start/UnityConfig.cs (hunk 9 above)
Comment (daveta) [reply csharp_webapi/2.EchoBot-With-Counter/App_Start/UnityConfig.cs]: Nice

**Comment by daveta** [comment]
@cleemullins : Thanks for the review, want to see if Drew can take a peek before we merge this one.

**Pull request merged**